### PR TITLE
SONARPY-200 Update S1110 implementation to only raise issue on double…

### PR DESCRIPTION
**Pull request by ivandalbosco**
… parens

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 1]
```diff
@@ -19,93 +19,142 @@
...
   @Override
-  public void init() {
-    subscribeTo(
-        PythonGrammar.TEST,
-        PythonGrammar.EXPR,
-        PythonGrammar.NOT_TEST
-    );
+  public void visitFile(AstNode astNode) {
+    leftParenthesisDataStack.clear();
+//    pairsByLevel.clear();
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 2]
```diff
@@ -19,93 +19,142 @@
...
+      if (previousTokenWasLeftParenthesis) {
+        leftParenthesisDataStack.peek().setContiguousLeftParenthesis(token);
       }
+      leftParenthesisDataStack.push(new LeftParenthensisData(token));
+      previousTokenWasLeftParenthesis = true;
+      previousTokenWasRightParenthesis = false;
+
+    } else if (token.getType() == RPARENTHESIS) {
+      checkRedundantParentheses(token);
+//      cleanUpPrevious();
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 3]
```diff
@@ -19,93 +19,142 @@
...
 @ActivatedByDefault
-public class UselessParenthesisCheck extends PythonCheck {
+public class UselessParenthesisCheck extends PythonCheck implements AstAndTokenVisitor {
+
   public static final String CHECK_KEY = "S1110";
 
+  private static final String MESSAGE = "Remove those useless parentheses.";
+
+  private Deque<LeftParenthensisData> leftParenthesisDataStack = new LinkedList<>();
+//  private LinkedList<LeftParenthensisData> leftParenthesisDataStack = new LinkedList<>();
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 4]
```diff
@@ -19,93 +19,142 @@
...
+
+    public void setRightParenthesis(Token rightParenthesis) {
+      this.rightParenthesis = rightParenthesis;
+    }
+
+    public boolean isFollowedByLeftParenthesis() {
+      return contiguousLeftParenthesis != null;
+    }
+
+//    public void setFollowedByLeftParenthesis(boolean isFollowedByLeftParenthesis) {
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 5]
```diff
@@ -19,93 +19,142 @@
...
   public static final String CHECK_KEY = "S1110";
 
+  private static final String MESSAGE = "Remove those useless parentheses.";
+
+  private Deque<LeftParenthensisData> leftParenthesisDataStack = new LinkedList<>();
+//  private LinkedList<LeftParenthensisData> leftParenthesisDataStack = new LinkedList<>();
+  
+  private LeftParenthensisData last;
+  
+  private Deque<Integer> pairsByLevel = new LinkedList<>();
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: MAJOR Remove this unused "pairsByLevel" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 6]
```diff
@@ -19,93 +19,142 @@
...
-    return parent.is(PythonGrammar.FOR_STMT) && parent.getFirstChild(PythonGrammar.EXPRLIST).getChildren(PythonGrammar.EXPR).size() == 1;
   }
+   */
 
-  private void checkAtom(AstNode atom, boolean ignoreTestNumber) {
-    if (violationCondition(atom, ignoreTestNumber)) {
-      addIssue(atom, "Remove those useless parentheses");
+  private void checkRedundantParentheses(Token rightParenthesis) {
+    LeftParenthensisData current = leftParenthesisDataStack.peek();
+//    LeftParenthensisData enclosing = leftParenthesisDataStack.get(leftParenthesisDataStack.size() - 2);
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 7]
```diff
@@ -19,93 +19,142 @@
...
+    public boolean isFollowedByLeftParenthesis() {
+      return contiguousLeftParenthesis != null;
+    }
+
+//    public void setFollowedByLeftParenthesis(boolean isFollowedByLeftParenthesis) {
+//      this.isFollowedByLeftParenthesis() = isFollowedByLeftParenthesis;
+//    }
+
+//    @Override
+//    public String toString() {
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 8]
```diff
@@ -19,93 +19,142 @@
...
 
-  private void checkAtom(AstNode atom, boolean ignoreTestNumber) {
-    if (violationCondition(atom, ignoreTestNumber)) {
-      addIssue(atom, "Remove those useless parentheses");
+  private void checkRedundantParentheses(Token rightParenthesis) {
+    LeftParenthensisData current = leftParenthesisDataStack.peek();
+//    LeftParenthensisData enclosing = leftParenthesisDataStack.get(leftParenthesisDataStack.size() - 2);
+    
+    if (previousTokenWasRightParenthesis && current.getContiguousLeftParenthesis() == last.getLeftParenthesis()) {  
+//      && enclosing.getContiguousLeftParenthesis() == current.getLeftParenthesis()) {
```
Comment (SonarTech) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 8 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 9]
```diff
@@ -19,93 +19,68 @@
...
 @ActivatedByDefault
 public class UselessParenthesisCheck extends PythonCheck {
+
   public static final String CHECK_KEY = "S1110";
 
+  private static final String MESSAGE = "Remove those useless parentheses.";
+
   @Override
   public void init() {
     subscribeTo(
```
Comment (pynicolas) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: Unit tests are still green if we replace this whole list with just ATOM and update visitNode accordingly.

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 10]
```diff
@@ -19,93 +19,68 @@
...
-        .children(PythonGrammar.TESTLIST_COMP)
-        .children(PythonGrammar.TEST)
-        .children(PythonGrammar.ATOM, PythonGrammar.COMPARISON)
-        .isNotEmpty();
-    if (hasUselessParenthesis) {
-      checkAtom(node.getFirstChild().getNextSibling(), false);
+    boolean hasParentheses = children.size() == 3 && children.get(0).is(PythonPunctuator.LPARENTHESIS);
+    if (hasParentheses) {
+      AstNode child1 = children.get(1);
+      if(child1.getChildren(PythonGrammar.TEST).size() == 1 && child1.getFirstChild(PythonPunctuator.COMMA) == null) {
```
Comment (pynicolas) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: Unit tests are still green when removing `&& child1.getFirstChild(PythonPunctuator.COMMA) == null`.

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 11]
```diff
@@ -19,93 +19,68 @@
...
+        AstNode test = child1.getChildren(PythonGrammar.TEST).get(0);
```
Comment (pynicolas) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: Why not `child1.getFirstChild(PythonGrammar.TEST)`?

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 12]
```diff
@@ -19,93 +19,68 @@
...
+        if (test.getChildren(PythonKeyword.IF).isEmpty()) {
```
Comment (pynicolas) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: `test.hasDirectChildren(PythonKeyword.IF)`

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 13]
```diff
@@ -19,93 +19,68 @@
...
+          AstNode testChild0 = test.getChildren().get(0);
```
Comment (pynicolas) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: `test.getFirstChild()`

File: its/ruling/src/test/resources/expected/python-S1110.json [hunk 14]
```diff
@@ -1,1749 +1,19 @@
...
-46,
-],
-'project:django-1.4/django/core/handlers/modpython.py':[
-130,
-],
-'project:django-1.4/django/core/management/color.py':[
-15,
-],
 'project:django-1.4/django/core/management/commands/dumpdata.py':[
 177,
```
Comment (pynicolas) [inline its/ruling/src/test/resources/expected/python-S1110.json]: This is a FP.

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 15]
```diff
@@ -19,93 +19,68 @@
...
 import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
-import org.sonar.sslr.ast.AstSelect;
+
+import com.sonar.sslr.api.AstNode;
 
 @Rule(
     key = UselessParenthesisCheck.CHECK_KEY,
     priority = Priority.MAJOR,
-    name = "Useless parentheses around expressions should be removed to prevent any misunderstanding",
+    name = "Redundant parentheses around expressions should be removed",
```
Comment (ivandalbosco) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: @pynicolas Thanks. Will be done in SONAR-192 (Rule API).

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java (hunk 11 above)
Comment (ivandalbosco) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: @pynicolas OK

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java (hunk 10 above)
Comment (ivandalbosco) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: @pynicolas [test](https://github.com/SonarSource/sonar-python/pull/90/commits/60ee50ac9e3f49f4f16be397ea0fd21135c05f01) added

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java (hunk 13 above)
Comment (ivandalbosco) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: @pynicolas OK

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java (hunk 12 above)
Comment (ivandalbosco) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: @pynicolas OK

File: its/ruling/src/test/resources/expected/python-S1110.json (hunk 14 above)
Comment (ivandalbosco) [inline its/ruling/src/test/resources/expected/python-S1110.json]: @pynicolas Thanks! Added handling of PythonGrammar.COMP_FOR.

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java [hunk 16]
```diff
@@ -19,93 +19,68 @@
...
-      if (selectedParent.size() == 1) {
-        AstNode parent = selectedParent.get(0);
-        if (isKeywordException(parent)) {
-          checkAtom(atom, true);
-          return;
-        }
-      }
-
-      checkAtom(atom, false);
+      checkAtom(node.getFirstChild().getNextSibling());
```
Comment (ivandalbosco) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: @pynicolas OK. No longer relevant.

File: python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java (hunk 9 above)
Comment (ivandalbosco) [inline python-checks/src/main/java/org/sonar/python/checks/UselessParenthesisCheck.java]: @pynicolas OK

**Pull request merged**